byzcoin/struct.go

```diff
@@ -1,12 +1,531 @@
+		}
+
+		// Each pair at this level is a bucket assigned to a skipchain
+		return b.ForEach(func(scid, v []byte) error {
+			scb := b.Bucket(scid)
+			if scb == nil {
+				return nil
+			}
+
+			var lastIndex int64
+
+			err := scb.ForEach(func(k, v []byte) error {
+				buf := bytes.NewBuffer(k[prefixLength+versionLength:])
+				var idx int64
+				err := binary.Read(buf, binary.BigEndian, &idx)
+				if err != nil {
+					return err
+				}
+
+				if idx > lastIndex {
+					lastIndex = idx
+				}
+
+				s.size += len(v)
+				return nil
+			})
+
+			indices[skipchain.SkipBlockID(scid).String()] = int(lastIndex + 1)
+			return err
+		})
+	})
+
+	return indices, err
+}
+
+// getBucket gets the bucket for the given skipchain
+func (s *stateChangeStorage) getBucket(tx *bolt.Tx, sid skipchain.SkipBlockID) *bolt.Bucket {
+	b := tx.Bucket(s.bucket)
+	if b == nil {
+		panic("Bucket has not been created. This is a programmer error.")
+	}
+
+	if tx.Writable() {
+		sbb, err := b.CreateBucketIfNotExists(sid)
+		if err != nil {
+			panic(err)
+		}
+
+		return sbb
+	}
+
+	return b.Bucket(sid)
+}
+
+// setMaxSize enables the cleaning of old state changes when the storage
+// size is above a given threshold. Note that the value is not strict.
+func (s *stateChangeStorage) setMaxSize(size int) {
+	s.maxSize = size
+}
+
+// setMaxNbrBlock enables the cleaning of state changes belonging
+// to blocks with an old index.
+func (s *stateChangeStorage) setMaxNbrBlock(nbr int) {
+	s.maxNbrBlock = nbr
+}
+
+// This will clean the oldest state changes when the total size
+// is above the maximum. It will remove elements until cleanThreshold of
+// the space is available.
+func (s *stateChangeStorage) cleanBySize() error {
+	if s.size < s.maxSize || s.maxSize == 0 {
+		// nothing to clean
+		return nil
+	}
+
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		// We make enough space to not have to do it everytime
+		// we append state changes
+		thres := int(float64(s.maxSize) * cleanThreshold)
+
+		for s.size > thres {
+			if len(s.sortedKeys) == 0 {
+				// This should never happen
+				return errors.New("Nothing can be cleaned, something is wrong with the storage implementation")
+			}
+
+			// Get the oldest version with the same instance ID
+			iid := s.sortedKeys[0].key[:prefixLength]
+
+			b := s.getBucket(tx, s.sortedKeys[0].scid)
+			c := b.Cursor()
+
+			k, v := c.Seek(iid)
+			if !bytes.HasPrefix(k, iid) {
+				// This should never occur
+				return errors.New("Missing key in the storage")
+			}
+
+			err := c.Delete()
+			if err != nil {
+				return err
+			}
+
+			s.size -= len(v)
+			k, v = c.Next()
+
+			if bytes.HasPrefix(k, iid) {
+				var sce StateChangeEntry
+				err := protobuf.Decode(v, &sce)
+				if err != nil {
+					return err
+				}
+				s.sortedKeys[0].time = sce.Timestamp
+				copy(s.sortedKeys[0].key, k)
+
+				sort.Sort(s.sortedKeys)
+			} else {
+				// if none, that means it was the last
+				s.sortedKeys = s.sortedKeys[1:]
```

Comment:
it's better to update the state after the transaction succeeds

Reply:
... but you're right on this one